processing/src/main/java/io/druid/segment/filter/JavaScriptFilter.java

```diff
@@ -51,34 +49,57 @@ public ImmutableBitmap getBitmapIndex(final BitmapIndexSelector selector)
   {
     final Context cx = Context.enter();
     try {
-      final Indexed<String> dimValues = selector.getDimensionValues(dimension);
       ImmutableBitmap bitmap;
-      if (dimValues == null) {
-        bitmap = selector.getBitmapFactory().makeEmptyImmutableBitmap();
-      } else {
-        bitmap = selector.getBitmapFactory().union(
-            FunctionalIterable.create(dimValues)
-                              .filter(
-                                  new Predicate<String>()
-                                  {
-                                    @Override
-                                    public boolean apply(@Nullable String input)
-                                    {
-                                      return predicate.applyInContext(cx, input);
-                                    }
-                                  }
-                              )
-                              .transform(
-                                  new com.google.common.base.Function<String, ImmutableBitmap>()
-                                  {
-                                    @Override
-                                    public ImmutableBitmap apply(@Nullable String input)
-                                    {
-                                      return selector.getBitmapIndex(dimension, input);
-                                    }
-                                  }
-                              )
-        );
+
+      boolean hasEmptyDimension = false;
+      Indexed<String> [] dimValuesList = new Indexed[dimension.length];
+      Iterator<String> [] dimValuesIterator = new Iterator[dimension.length];
+      String[] currentDim = new String[dimension.length];
+      for(int idx = 0; idx < dimension.length; idx++) {
+        dimValuesList[idx] = selector.getDimensionValues(dimension[idx]);
+        if (dimValuesList[idx].size() == 0) {
+          hasEmptyDimension = true;
+          break;
+        }
+        dimValuesIterator[idx] = dimValuesList[idx].iterator();
+        if (idx != 0) {
+          currentDim[idx] = dimValuesIterator[idx].next();
+        }
+      }
+
+      bitmap = selector.getBitmapFactory().makeEmptyImmutableBitmap();
+      if (!hasEmptyDimension) {
```

Comment:
@sirpkt If a dimension didn't have any values at all, it wouldn't be persisted (null is only in the dictionary if the dim has null + at least 1 non-null value, otherwise the column is just skipped), so getDimensionValues() would return null in that case.

However, other dimensions could still have values; suppose the user defined a schema with dimA and dimB; it's possible that some subset of segments has no value for dimB (i.e., no rows with dimB were seen during the intervals covered by those segments).

So if the user has a filter like (dimA == null && dimB == 'X), seems like dimB values still need to be checked, if a segment without dimA is encountered